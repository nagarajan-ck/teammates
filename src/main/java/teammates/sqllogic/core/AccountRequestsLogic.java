package teammates.sqllogic.core;

import java.util.List;

import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.common.exception.SearchServiceException;
import teammates.common.util.HibernateUtil;
import teammates.storage.sqlapi.AccountRequestsDb;
import teammates.storage.sqlentity.AccountRequest;
import teammates.storage.sqlsearch.AccountRequestSearchManager;

/**
 * Handles operations related to account requests.
 *
 * @see AccountRequest
 * @see AccountRequestsDb
 */
public final class AccountRequestsLogic {

    private static final AccountRequestsLogic instance = new AccountRequestsLogic();

    private AccountRequestsDb accountRequestDb;

    private AccountRequestsLogic() {
        // prevent notification
    }

    public static AccountRequestsLogic inst() {
        return instance;
    }

    /**
     * Initialise dependencies for {@code AccountRequestLogic} object.
     */
    public void initLogicDependencies(AccountRequestsDb accountRequestDb) {
        this.accountRequestDb = accountRequestDb;
    }

    private AccountRequestSearchManager getSearchManager() {
        return accountRequestDb.getSearchManager();
    }

    /**
     * Creates or updates search document for the given account request.
     */
    public void putDocument(AccountRequest accountRequest) throws SearchServiceException {
        getSearchManager().putDocument(accountRequest);
    }

    /**
     * Creates an account request.
     */
    public AccountRequest createAccountRequest(AccountRequest accountRequest)
            throws InvalidParametersException, EntityAlreadyExistsException {
        return accountRequestDb.createAccountRequest(accountRequest);
    }

    /**
     * Creates an account request.
     */
    public AccountRequest createAccountRequest(String name, String email, String institute)
            throws InvalidParametersException, EntityAlreadyExistsException {
        AccountRequest toCreate = new AccountRequest(email, name, institute);

        return accountRequestDb.createAccountRequest(toCreate);
    }

    /**
     * Gets account request associated with the {@code email} and {@code institute}.
     */
    public AccountRequest getAccountRequest(String email, String institute) {

        return accountRequestDb.getAccountRequest(email, institute);
    }

    /**
     * Updates an account request.
     */
    public AccountRequest updateAccountRequest(AccountRequest accountRequest)
            throws InvalidParametersException, EntityDoesNotExistException {
        return accountRequestDb.updateAccountRequest(accountRequest);
    }

    /**
     * Gets account request associated with the {@code regkey}.
     */
    public AccountRequest getAccountRequestByRegistrationKey(String regkey) {
        return accountRequestDb.getAccountRequestByRegistrationKey(regkey);
    }

    /**
     * Creates/resets the account request with the given email and institute such that it is not registered.
     */
    public AccountRequest resetAccountRequest(String email, String institute)
            throws EntityDoesNotExistException, InvalidParametersException {
        AccountRequest accountRequest = accountRequestDb.getAccountRequest(email, institute);

        if (accountRequest == null) {
            throw new EntityDoesNotExistException("Failed to reset since AccountRequest with "
                + "the given email and institute cannot be found.");
        }
        accountRequest.setRegisteredAt(null);

        return accountRequestDb.updateAccountRequest(accountRequest);
    }

    /**
     * Deletes account request associated with the {@code email} and {@code institute}.
     *
     * <p>Fails silently if no account requests with the given email and institute to delete can be found.</p>
     *
     */
    public void deleteAccountRequest(String email, String institute) {
        AccountRequest toDelete = accountRequestDb.getAccountRequest(email, institute);

        accountRequestDb.deleteAccountRequest(toDelete);
    }

    /**
     * Searches for account requests in the whole system.
     *
     * @return A list of {@link AccountRequest} or {@code null} if no match found.
     */
    public List<AccountRequest> searchAccountRequestsInWholeSystem(String queryString)
            throws SearchServiceException {
        return accountRequestDb.searchAccountRequestsInWholeSystem(queryString);
    }

    /**
     * Creates an or gets an account request.
     */
    public AccountRequest createOrGetAccountRequestWithTransaction(String name, String email, String institute)
            throws InvalidParametersException {
        AccountRequest toCreate = new AccountRequest(email, name, institute);
        HibernateUtil.beginTransaction();
        AccountRequest accountRequest;
        try {
            accountRequest = accountRequestDb.createAccountRequest(toCreate);
            HibernateUtil.commitTransaction();
        } catch (InvalidParametersException ipe) {
            HibernateUtil.rollbackTransaction();
            throw new InvalidParametersException(ipe);
        } catch (EntityAlreadyExistsException eaee) {
            // Use existing account request
            accountRequest = getAccountRequest(email, institute);
            HibernateUtil.commitTransaction();
        }
        return accountRequest;
    }
}
