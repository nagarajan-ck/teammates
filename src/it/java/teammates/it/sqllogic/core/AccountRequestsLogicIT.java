package teammates.it.sqllogic.core;

import java.time.Instant;

import org.testng.annotations.Test;

import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.AccountRequestsLogic;
import teammates.storage.sqlapi.AccountRequestsDb;
import teammates.storage.sqlentity.AccountRequest;

/**
 * SUT: {@link AccountRequestsLogic}.
 */
public class AccountRequestsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private AccountRequestsLogic accountRequestsLogic = AccountRequestsLogic.inst();

    @Test
    public void testResetAccountRequest()
            throws EntityAlreadyExistsException, InvalidParametersException, EntityDoesNotExistException {

        ______TS("success: create account request and update registeredAt field");

        String name = "name lee";
        String email = "email@gmail.com";
        String institute = "institute";

        AccountRequest toReset = accountRequestsLogic.createAccountRequest(name, email, institute);
        AccountRequestsDb accountRequestsDb = AccountRequestsDb.inst();

        toReset.setRegisteredAt(Instant.now());
        toReset = accountRequestsDb.getAccountRequest(email, institute);

        assertNotNull(toReset);
        assertNotNull(toReset.getRegisteredAt());

        ______TS("success: reset account request that already exists");

        AccountRequest resetted = accountRequestsLogic.resetAccountRequest(email, institute);

        assertNull(resetted.getRegisteredAt());

        ______TS("success: test delete account request");

        accountRequestsLogic.deleteAccountRequest(email, institute);

        assertNull(accountRequestsLogic.getAccountRequest(email, institute));

        ______TS("failure: reset account request that does not exist");

        assertThrows(EntityDoesNotExistException.class,
                () -> accountRequestsLogic.resetAccountRequest(name, institute));
    }
}
