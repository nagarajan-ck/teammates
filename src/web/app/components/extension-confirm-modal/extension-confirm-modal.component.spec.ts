import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
import { TimezoneService } from '../../../services/timezone.service';
import {
  FeedbackSession, FeedbackSessionPublishStatus, FeedbackSessionSubmissionStatus,
} from '../../../types/api-output';
import { ResponseVisibleSetting, SessionVisibleSetting } from '../../../types/api-request';
import {
  StudentExtensionTableColumnModel,
} from '../../pages-instructor/instructor-session-individual-extension-page/extension-table-column-model';
import { ExtensionConfirmModalComponent } from './extension-confirm-modal.component';
import { ExtensionConfirmModalModule } from './extension-confirm-modal.module';

describe('ExtensionConfirmModalComponent', () => {
  const testFeedbackSession: FeedbackSession = {
    courseId: 'testId1',
    timeZone: 'Asia/Singapore',
    feedbackSessionName: 'Test Session',
    instructions: 'Instructions',
    submissionStartTimestamp: 1000000000000,
    submissionEndTimestamp: 1500000000000,
    gracePeriod: 0,
    sessionVisibleSetting: SessionVisibleSetting.AT_OPEN,
    responseVisibleSetting: ResponseVisibleSetting.AT_VISIBLE,
    submissionStatus: FeedbackSessionSubmissionStatus.OPEN,
    publishStatus: FeedbackSessionPublishStatus.PUBLISHED,
    isClosingEmailEnabled: true,
    isPublishedEmailEnabled: true,
    createdAtTimestamp: 0,
    studentDeadlines: {},
    instructorDeadlines: {},
  };

  const studentModel1: StudentExtensionTableColumnModel = {
    sectionName: 'Test Section 1',
    teamName: 'Test Section 1',
    name: 'Test Student 1',
    email: 'testStudent1@gmail.com',
    extensionDeadline: 1500000000000,
    hasExtension: false,
    isSelected: true,
  };
  const studentModel2: StudentExtensionTableColumnModel = {
    sectionName: 'Test Section 2',
    teamName: 'Test Section 2',
    name: 'Test Student 2',
    email: 'testStudent2@gmail.com',
    extensionDeadline: 1510000000000,
    hasExtension: true,
    isSelected: true,
  };
  const studentModel3: StudentExtensionTableColumnModel = {
    sectionName: 'Test Section 3',
    teamName: 'Test Section 3',
    name: 'Test Student 3',
    email: 'testStudent3@gmail.com',
    extensionDeadline: 1510000000000,
    hasExtension: true,
    isSelected: true,
  };

  const testTimeString = '5 Apr 2000 2:00:00';

  let component: ExtensionConfirmModalComponent;
  let fixture: ComponentFixture<ExtensionConfirmModalComponent>;
  let timeZoneService: TimezoneService;

  beforeEach(
    waitForAsync(() => {
      TestBed.configureTestingModule({
        imports: [HttpClientTestingModule, ExtensionConfirmModalModule],
        providers: [NgbActiveModal],
      }).compileComponents();
    }),
  );

  beforeEach(() => {
    fixture = TestBed.createComponent(ExtensionConfirmModalComponent);
    timeZoneService = TestBed.inject(TimezoneService);
    jest.spyOn(timeZoneService, 'formatToString').mockReturnValue(testTimeString);
    component = fixture.componentInstance;
    component.feedbackSessionTimeZone = 'Asia/Singapore';
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with the extended students and instructors', () => {
    component.studentData = [studentModel1, studentModel2, studentModel3];
    component.extensionTimestamp = testFeedbackSession.submissionEndTimestamp;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
