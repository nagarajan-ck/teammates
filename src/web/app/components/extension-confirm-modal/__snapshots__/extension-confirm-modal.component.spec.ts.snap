// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExtensionConfirmModalComponent should snap with the extended students and instructors 1`] = `
<tm-extension-confirm-modal
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  __ngContext__={[Function Number]}
  activeModal={[Function NgbActiveModal]}
  confirmExtensionCallbackEvent={[Function EventEmitter_]}
  dateDetailPipe={[Function FormatDateDetailPipe]}
  extensionTimestamp={[Function Number]}
  feedbackSessionTimeZone={[Function String]}
  headerColorScheme={[Function Number]}
  instructorColumnsData={[Function Array]}
  instructorRoleNamePipe={[Function InstructorRoleNamePipe]}
  instructorRowsData={[Function Array]}
  isNotifyDeadlines="false"
  isSubmitting="false"
  modalType="0"
  selectedInstructors={[Function Array]}
  selectedStudents={[Function Array]}
  sortInstructorListEvent={[Function EventEmitter_]}
  sortInstructorOrder="0"
  sortInstructorsBy={[Function Number]}
  sortStudentListEvent={[Function EventEmitter_]}
  sortStudentOrder="0"
  sortStudentsBy={[Function Number]}
  studentColumnsData={[Function Array]}
  studentRowsData={[Function Array]}
  tableComparatorService={[Function TableComparatorService]}
  timeZoneService={[Function TimezoneService]}
>
  <fieldset>
    <div
      class="modal-header alert-primary"
    >
      <h5
        class="modal-title"
      >
        Confirm extension of feedback session
      </h5>
      <button
        class="btn-close"
        type="button"
      />
    </div>
    <div
      class="modal-body"
    >
      <div>
         The new deadline of 
        <b>
          5 Apr 2000 2:00:00
        </b>
         will be applied to the following 
        <b>
          3
        </b>
         student(s) and 
        <b>
          0
        </b>
         instructor(s): 
      </div>
      <div>
        <h1>
          Students
        </h1>
        <tm-sortable-table>
          <table
            class="table table-bordered table-striped"
          >
            <thead>
              <tr>
                <th
                  aria-sort="none"
                  class="student-sort-by-section sortable-header"
                  style="text-align: start;"
                >
                  <button
                    class="d-lg-flex align-items-center gap-lg-1 d-inline"
                  >
                    <span
                      container="body"
                    >
                      Section
                    </span>
                    <span
                      aria-hidden="true"
                      class="fa-stack"
                    >
                      <i
                        class="fas fa-sort"
                      />
                    </span>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  class="student-sort-by-team sortable-header"
                  style="text-align: start;"
                >
                  <button
                    class="d-lg-flex align-items-center gap-lg-1 d-inline"
                  >
                    <span
                      container="body"
                    >
                      Team
                    </span>
                    <span
                      aria-hidden="true"
                      class="fa-stack"
                    >
                      <i
                        class="fas fa-sort"
                      />
                    </span>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  class="student-sort-by-name sortable-header"
                  style="text-align: start;"
                >
                  <button
                    class="d-lg-flex align-items-center gap-lg-1 d-inline"
                  >
                    <span
                      container="body"
                    >
                      Name
                    </span>
                    <span
                      aria-hidden="true"
                      class="fa-stack"
                    >
                      <i
                        class="fas fa-sort"
                      />
                    </span>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  class="student-sort-by-email sortable-header"
                  style="text-align: start;"
                >
                  <button
                    class="d-lg-flex align-items-center gap-lg-1 d-inline"
                  >
                    <span
                      container="body"
                    >
                      Email
                    </span>
                    <span
                      aria-hidden="true"
                      class="fa-stack"
                    >
                      <i
                        class="fas fa-sort"
                      />
                    </span>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  class="student-sort-by-deadline sortable-header"
                  style="text-align: start;"
                >
                  <button
                    class="d-lg-flex align-items-center gap-lg-1 d-inline"
                  >
                    <span
                      container="body"
                    >
                      Original Deadline
                    </span>
                    <span
                      aria-hidden="true"
                      class="fa-stack"
                    >
                      <i
                        class="fas fa-sort"
                      />
                    </span>
                  </button>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  Test Section 1
                </td>
                <td>
                  Test Section 1
                </td>
                <td>
                  Test Student 1
                </td>
                <td>
                  testStudent1@gmail.com
                </td>
                <td>
                  <span>
                    5 Apr 2000 2:00:00
                  </span>
                </td>
              </tr>
              <tr>
                <td>
                  Test Section 2
                </td>
                <td>
                  Test Section 2
                </td>
                <td>
                  Test Student 2
                </td>
                <td>
                  testStudent2@gmail.com
                </td>
                <td>
                  <span>
                    5 Apr 2000 2:00:00
                  </span>
                </td>
              </tr>
              <tr>
                <td>
                  Test Section 3
                </td>
                <td>
                  Test Section 3
                </td>
                <td>
                  Test Student 3
                </td>
                <td>
                  testStudent3@gmail.com
                </td>
                <td>
                  <span>
                    5 Apr 2000 2:00:00
                  </span>
                </td>
              </tr>
            </tbody>
          </table>
        </tm-sortable-table>
      </div>
      <div
        class="form-check"
      >
        <input
          class="form-check-input ng-untouched ng-pristine ng-valid"
          id="flexCheckChecked"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="flexCheckChecked"
        >
           Notify affected students and/or instructors by email about this extension 
        </label>
      </div>
    </div>
    <br />
    <div
      class="modal-footer py-1 px-3"
    >
      <button
        class="btn btn-light"
        type="button"
      >
        No, cancel the Operation
      </button>
      <button
        class="btn btn-success modal-btn-ok"
        type="button"
      >
        Confirm
      </button>
    </div>
  </fieldset>
</tm-extension-confirm-modal>
`;
